File: LeetCodePractice/src/bst/FindDuplicateSubtrees.java
package bst;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/*
Given the root of a binary tree, return all duplicate subtrees.
For each kind of duplicate subtrees, you only need to return the root node of any one of them.
Two trees are duplicate if they have the same structure with the same node values.

Example 1:
Input: root = [1,2,3,4,null,2,4,null,null,4]
Output: [[2,4],[4]]

Example 2:
Input: root = [2,1,1]
Output: [[1]]

Example 3:
Input: root = [2,2,2,3,null,3,null]
Output: [[2,3],[3]]

Constraints:
The number of the nodes in the tree will be in the range [1, 10^4]
-200 <= Node.val <= 200

Analysis:
TC O(N)
 */
public class FindDuplicateSubtrees {
    int id = 1;

    public List<TreeNode> findDuplicateSubtrees(TreeNode root) {
        Map<String, Integer> serialToId = new HashMap<>();
        Map<Integer, Integer> idToCnt = new HashMap<>();
        List<TreeNode> res = new ArrayList<>();
        postOrder(root, serialToId, idToCnt, res);
        return res;
    }

    private int postOrder(TreeNode root, Map<String, Integer> serialToId, Map<Integer, Integer> idToCnt, List<TreeNode> res) {
        if (root == null) {
            return 0;
        }
        int leftId = postOrder(root.left, serialToId, idToCnt, res);
        int rightId = postOrder(root.right, serialToId, idToCnt, res);
        String serial = leftId + "," + root.val + "," + rightId;
        int serialId = serialToId.getOrDefault(serial, id);
        if (serialId == id) {
            id++;
        }
        serialToId.put(serial, serialId);
        idToCnt.put(serialId, idToCnt.getOrDefault(serialId, 0) + 1);
        if (idToCnt.get(serialId) == 2) {
            res.add(root);
        }
        return serialId;
    }

}

File: LeetCodePython/FindDuplicateSubtrees.py
"""
Given the root of a binary tree, return all duplicate subtrees.

For each kind of duplicate subtrees, you only need to return the root node of any one of them.

Two trees are duplicate if they have the same structure with the same node values.

Example 1:
Input: root = [1,2,3,4,null,2,4,null,null,4]
Output: [[2,4],[4]]

Example 2:
Input: root = [2,1,1]
Output: [[1]]

Example 3:
Input: root = [2,2,2,3,null,3,null]
Output: [[2,3],[3]]

Constraints:
The number of the nodes in the tree will be in the range [1, 5000]
-200 <= Node.val <= 200

analysis:
pre-order serialization
TC: O(N)
"""
from collections import defaultdict
from typing import Optional, List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class FindDuplicateSubtrees:
    def findDuplicateSubtrees(self, root: Optional[TreeNode]) -> List[Optional[TreeNode]]:
        res = []
        serial_cnt = defaultdict(int)

        def dfs(cur, serial):
            if not cur:
                return '#'
            left_serial, right_serial = dfs(cur.left, serial), dfs(cur.right, serial)
            serial += ','.join([str(cur.val), left_serial, right_serial])
            serial_cnt[serial] += 1
            if serial_cnt[serial] == 2:
                res.append(cur)
            return serial

        dfs(root, '')
        return res
